### ISSUE TITLE
React Native Project keeps loading infinitely & heats up CPU

### ISSUE THREAD

**Issue by deadcoder0904** [pnpm/pnpm#1252]
### Code to reproduce the issue:

https://github.com/deadcoder0904/react-native-darkmode-list

### Expected behavior:

Code should run properly

### Actual behavior:

Code runs infinitely. I first tried simple `pnpm install` but it raised an error about some babel plugin not installed. So I set `.npmrc` file to `shamefully-flatten=true` & ran `pnpm install` & it went well.

Then I typed `pnpm start` & it keeps on running infinitely while heating my CPU. Same thing with `yarn` runs in few seconds like 10-15 & with `pnpm` I waited 2 minutes until it got very much heated that I closed the running process.

### Additional information:

 - `node -v` prints: 10.4.1
 - Windows, OS X, or Linux?: OS X

**Comment by zkochan** [comment]
TBH I have no experience with React Native but I remember @vjpr mentioning that it has issues with symlinks (I think @pgonzal also mentioned it).

Can we start communication with the React Native about these issues? I am almost sure that we can't do anything on our (pnpm) end.

**Comment by deadcoder0904** [comment]
I think I might've used it for a couple of projects but not sure. Since I often jump between `pnpm` & `yarn` when something doesn't work. But, yeah if its a known issue I'll use something else with React Native for now :)

**Comment by rodrigobdz** [comment]
In my opinion, this should be in the README. I just migrated to pnpm and first thing I discover while trying to run my React Native project is that it doesn't work.

**Comment by pgonzal** [comment]
A while ago someone explained to me that there might be a cultural difference behind this issue: Typical NodeJS developers expect their code to run on any OS, with any package manager, and without installing any other software. Whereas native app developers often must install huge SDKs with multiple prerequisites and special device drivers, for each device that they target. For iOS you even need a specific development machine. In that context, the idea that you must use a certain certain package manager doesn't seem like a big concern. It's just one more "hoop" you have to jump through to make an app. The theory was that this is why there is relatively little effort invested in fixing the React Native symlink handling bugs, whereas other NodeJS packages get fixed quickly.

Not sure if it's true. (I haven't used React Native myself beyond trying the demos. But we hear about this issue occasionally from people who are wanting to enable PNPM for their Rush repositories.)

**Comment by zkochan** [comment]
Are there any updates with this issue? I think they will want to support Yarn Plug'n'Play, so they might make React Native more flexible

**Comment by likern** [comment]
> @zkochan  Are there any updates with this issue? I think they will want to support Yarn Plug'n'Play, so they might make React Native more flexible

To my knowledge the problem is not in yarn per se or package manager at all.
It's the problem of package bundler (like `webpack`).
React Native uses Facebook's **Metro** bundler https://github.com/facebook/metro which converts all javascript files into one file. This bundler doesn't support symlinks.

This issue is the root of the problems: https://github.com/facebook/metro/issues/1

I see the following approaches to overcome this limitation:
1. If we could use **only hard links** that should solve the problem
2. utilize https://github.com/wix/wml which runs watcher and everytime files are changed copy over content directly (doesn't create symlinks)

**Comment by octogonz** [comment]
Why not just fix Metro?

**Comment by likern** [comment]
> Why not just fix Metro?

**It should, but...** but will not be fixed anytime soon, at least in foreseeable future.
I've made some research and want to bring in some additional context about this.

### Start of the story
Original issue  https://github.com/facebook/metro/issues/1 was created **1252 days** or **3 years 5 months 6 days** ago.
**1 year 10 months 15 days ago** @aleclarson provided pull request https://github.com/facebook/metro/pull/257 fixing this issue.
Metro team first required to merge _symlinks support in Jest_ (not sure how they are related, might be this breaks Jest tests).

Another pull request https://github.com/facebook/jest/pull/6993 to fix issue in Jest was provided by @aleclarson **1 year 10 months 15 days ago** (and later, **1 year 7 month 7 days ago**, was provided updated pull request https://github.com/facebook/jest/pull/7549 superseding original).

It turned out that **Jest relies on Watchman** (which watches file changes) which also **doesn't have symlinks support**.
https://github.com/facebook/jest/pull/7549#issuecomment-475741010
> Basically, for performance, we rely on Watchman telling us what files are changed. Watchman has no support for symlinks (facebook/watchman#105). That means if files are updated from a symlink, the haste map won't update it.

Jest developers **refused to merge PR** requesting _"transparent symlink"_ implementation (don't know what that means) instead of resolving (cloning content) symlinks.
https://github.com/facebook/jest/pull/7549#issuecomment-475781263
> The reason I'm hesitant about this is because ...
> ... if it's possible to implement this in a way that hides symlinks from the rest of the system

**Watchman** also has issue about symlinks support https://github.com/facebook/watchman/issues/105. It was created **3 years 8 months 26 days ago**.
https://github.com/facebook/watchman/issues/105#issuecomment-105270042

**Watchman** developers said in comments that adding symlinks support
> ... would be tremendously complex and still be error prone. As a result, it is unlikely that we'll ever add support for resolving and tracking symlink targets

### Final thoughts
1. For Facebook developers adding symlinks is a rocket science.
2. Reviewing pull requests (to look at first glance on PR) **might take more than two months**
3. All Pull Requests are stale and **not moving forward at all** (there is even no more comments from devs)
4. Metro waits Jest pull request merge (+735, −233 sloc), Jest **waits on new implementation**, Watchman **refused to add support at all by themselves**
5. **No symlinks support anytime soon**. _In my estimation in period of 3-5 years at least_.

**Comment by likern** [comment]
So if pnpm would want to **support React Native** and could solve that problem anyhow by themselves (by hacks / workarounds / etc) - I beg to do that.

I think the most _straightforward_ and _bulletproof_ solution would be to add ability to use **original npm resolution** - 
1. Resolve all symlinks
2. Don't create symlinks at all and have deep nested original structure (yes, with package content duplication)

But **this is the most simple solution** I think. It might be turned on in some config (package.json) only on React Native projects
_at the last resort_. **Thus no symlinks will be created at all**.

But this will bring in all the benefits of pnpm to React Native world:
1. Promised **strictness**
2. Works with workspaces
3. Merely substitutues lerna
4. Unified experience
5. **Can fallback to symlinks** whenever they will be supported by Metro

**Comment by likern** [comment]
Eventually on my machine all `node_modules` take up only
```bash
find . -type d -name node_modules |  grep -v node_modules.*node_modules | tr '\n' '\0' |  xargs -0 du -sch
30G total
```
we can deal with it.

**Comment by zkochan** [comment]
What you suggest would require months of development. Just to support a community that doesn't support us.

The only solution for this would be to use Yarn's Plug'n'Play. At some point, we will probably support it. And with that, symlinks are not needed.

**Comment by aleclarson** [comment]
@likern You could try using [Haul](https://github.com/callstack/haul), an alternative to Metro. They claim to support symlinks. I've not tried it myself, as I'm using the PRs that are stalled. Another option, I could generate some git patches from my PRs to both `jest` and `metro`, which you can apply with [patch-package](https://github.com/ds300/patch-package).

**Comment by likern** [comment]
> @likern You could try using [Haul](https://github.com/callstack/haul), an alternative to Metro. They claim to support symlinks. I've not tried it myself, as I'm using the PRs that are stalled. Another option, I could generate some git patches from my PRs to both `jest` and `metro`, which you can apply with [patch-package](https://github.com/ds300/patch-package). But then you would be stuck on the version of Metro that I'm using (the Jest changes are up-to-date, I think).

Yes, but they do not support Fast Refresh. Which for me is more important feature.

**Comment by likern** [comment]
> ... I'm using the PRs that are stalled.

Might be you could publish your fork of Metro with  patches applied?
So other devs could just swap original Metro to patched version?

**Comment by sammarks** [comment]
@aleclarson do you have the time to generate those patches for `patch-package`? That would be tremendously useful while we wait for the PRs to be merged!

**Comment by zkochan** [comment]
Does React native work with Yarn Plug'n'Play?
If yes, you can now create a `.pnp.js` file from a pnpm lockfile. Also, we plan to integrate PnP more to pnpm: https://github.com/pnpm/pnpm/issues/2902

**Comment by merceyz** [comment]
It does not, they rely on the node_modules layout so much Yarn v1 had to implement nohoist for it to even work. v2 will have something similar, but better, in the next release https://github.com/yarnpkg/berry/pull/1819

**Comment by vjpr** [comment]
I spent a lot of time investigating this issue over the years. Pnpm works fine with React Native. There is a small patch to Métro needed and then you just use a custom module resolver - webpack's 'enhanced-resolve'.

There were some small patches in jest waiting to merge last time I checked a couple months ago. 

When I have some free time I will create a robust example project.

**Comment by billouboq** [comment]
@vjpr Do you have a snippet of your code to make it work with pnpm ?
That would be awesome ! 👏

**Comment by ShivamJoker** [comment]
@vjpr it's been 2 months please post something

**Comment by vjpr** [comment]
Will post tomorrow!

**Comment by marudy** [comment]
@vjpr if you can post that buddy, you will save our lives :) I'm super excited to try out yarn berry or pnpm in order to improve `npm install` performance but it's unfortunate that none does that yet :/

**Comment by marudy** [comment]
@vjpr sorry for spamming you man but it seems you have done the work already and i'm trying to gather the pieces to make pnpm to work for my React Native project as well. This is the metro patch you are referring to I suppose: https://github.com/pnpm/pnpm/issues/1051 correct ?

**Comment by vjpr** [comment]
Diving back into this now. Want to try and solve it for good because it takes a lot of mental effort to wrap my head around all the pieces each time I return to it. Stay tuned.

**Comment by vjpr** [comment]
@marudy @ShivamJoker Please see https://github.com/pnpm/pnpm/issues/3010

Working example: https://github.com/vjpr/pnpm-expo-example

**Comment by airtonix** [comment]
> Eventually on my machine all `node_modules` take up only
> 
> ```shell
> find . -type d -name node_modules |  grep -v node_modules.*node_modules | tr '\n' '\0' |  xargs -0 du -sch
> 30G total
> ```
> 
> we can deal with it.

not in CICD we can't.

**Comment by zkochan** [comment]
For anyone finding this issue through search. The easiest way to make React Native work with pnpm is to use the `node-linker=hoisted` setting. With this setting pnpm will create a flat node_modules without using symlinks. Related docs: https://pnpm.io/npmrc#node-linker

**Comment by byCedric** [comment]
In addition to the `node-linker=hoisted`, I want to mention the Metro config outlined in the Expo docs "[Working with monorepos](https://docs.expo.dev/guides/monorepos/#modify-the-metro-config)". 

I tried to explain most of the issues you probably will run into with monorepos, React Native, and/or Metro. With that context, you should be able to solve it in your preferred way (e.g. rnx-kit, or just adopt the same Metro config properties). There is [also this repository that uses Expo, pnpm, and turborepo](https://github.com/byCedric/expo-monorepo-example).
